### PR TITLE
Improve performance of DateTime searches in Cosmos DB

### PR DESCRIPTION
**Pull request by johnstairs**
## Description
Searches like `/Patient?birthdate=1948-11-22` currently perform poorly with the Cosmos DB provider. In this change, we are making the following change to the query we generate this this kind of search:
image

Today, the database is forced to intersect all patients where the birthdate's start is > queryStart and birthdate's end is < queryEnd, but because there is no correlation between these fields, this results in all patients being processed to perform the intersection. 

The solution is to add an upper bound to start, the same as the end time. 

Since we already do this for the SQL implementation, we are reusing the same visitor.

This makes a major difference. For a database with 44K Synthea patients, here are some before and after numbers, searching for the first ~10 results:

|                                          | Before RUs                                               | After RUs |
|------------------------------------------|----------------------------------------------------------|-----------|
| Observation?date=2011-08-17              | 143283                                                   | 18        |
| /Patient?birthdate=1948-11-22            | 1911                                                     | 15        |
| Observation?date=2007-08-16 (no matches) | Many empty pages with continuation and 5-10K RU per page | 7.85      |

## Related issues
Addresses #1690 

## Testing
Added unit tests and manual testing against a large-ish database

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch

### CODE REVIEW

**Review by brendankowitz — APPROVED** [review]

**Comment by LTA-Thinking** [comment]
Is it worth also adding the constraint that end time has to be greater than or equal to parameter 3?

File: src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/DateTimeEqualityRewriter.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+namespace Microsoft.Health.Fhir.Core.Features.Search.Expressions
+{
+    /// <summary>
+    /// Rewrites (And (FieldGreaterThanOrEqual DateTimeStart x) (FieldLessThanOrEqual DateTimeEnd y)) to
+    /// (And (FieldGreaterThanOrEqual DateTimeStart x) (FieldLessThanOrEqual DateTimeEnd y) (FieldLessThanOrEqual DateTimeStart y)).
+    /// It looks specifically for this pattern because it is what we emit in the core layer for an equals search over a date parameter.
+    /// This rewriting constrains the range scan over the index (DateTimeStart, DateTimeEnd).
+    /// </summary>
+    internal class DateTimeEqualityRewriter : ExpressionRewriterWithInitialContext<object>
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/DateTimeEqualityRewriter.cs:18]: >DateTimeEqualityRewriter [](start = 19, length = 24)

nit: is there are reason to move it out of `Visitors` folder? All other rewriters are remain in that folder. Also it's a bit tricky to see changes in file.

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Search/FhirCosmosSearchService.cs [hunk 2]
```diff
@@ -57,6 +57,12 @@ public class FhirCosmosSearchService : SearchService
             SearchOptions searchOptions,
             CancellationToken cancellationToken)
         {
+            // we're going to mutate searchOptions, so clone it first so the caller of this method does not see the changes.
+            searchOptions = searchOptions.Clone();
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Search/FhirCosmosSearchService.cs:61]: This looks like a shallow clone.  If so any reference properties would be shared between the original object and the clone.  Do we alter the expression object that is passed to the Visitor, or de we generate a new object which is assigned to the clone?

Comment (johnstairs) [reply src/Microsoft.Health.Fhir.CosmosDb/Features/Search/FhirCosmosSearchService.cs:61]: The expressions are all immutable so this should be safe

File: src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/DateTimeEqualityRewriter.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                return expression;
+            }
+
+            List<Expression> newExpressions = null;
+            int i = 0;
+            for (; i < expression.Expressions.Count - 1; i++)
+            {
+                switch (MatchPattern(expression.Expressions[i], expression.Expressions[i + 1]))
+                {
+                    case ({ } low, { } high):
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/DateTimeEqualityRewriter.cs:47]: >{ } low, { } high [](start = 26, length = 17)

I never saw this syntax before. What these curly braces do?

Comment (johnstairs) [reply src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/DateTimeEqualityRewriter.cs:47]: It matches non-null values. Alternative would be `(BinaryExpression low, BinaryExpression high)`. But `(var low, var high)` matches null values which is pretty surprising design decision IMO.

**Comment by johnstairs** [comment]
> Is it worth also adding the constraint that end time has to be greater than or equal to parameter 3?

@LTA-Thinking. unless the data is invalid in the database, a query looking for endtime < starttime would return no results. Without this optimization, it would require a table scan to realize that. With this, it will find that out faster :)

**Pull request merged**